### DISCUSSION TITLE
How to take grain size into consideration in uniaxial tension

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#19199] [Q&A Modules: Solid mechanics] [answered]
Dear Moose expert or users,

Could anyone share your ideas or thoughts about how to take grain/particle size into consideration in power-law model (https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html) for uniaxial tension simulation and look at how the effect of grain/particle size on the stress-strain curve? Thank you.

**Comment by sapitts** [comment, marked as answer]
Hi @xchengood,
You may like to consider a Hall-Petch type term (for grain size influence) or an impenetrable barrier type model (for embedded particles) as additional hardening terms. 

In a class that inherits from `CrystalPlasticityKalidindiUpdate`, I'd recommend either overwriting the method `calculateStateVariableEvolutionRateComponent` or adding a new constant hardening contribution term that is called in the stateful property initialization method. An example of this second approach is used in the `initQpStatefulProperties` method of [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.html).

Hope this helps,
Stephanie

**Reply by xchengood** [reply]
> Hi @xchengood, You may like to consider a Hall-Petch type term (for grain size influence) or an impenetrable barrier type model (for embedded particles) as additional hardening terms.
> 
> In a class that inherits from `CrystalPlasticityKalidindiUpdate`, I'd recommend either overwriting the method `calculateStateVariableEvolutionRateComponent` or adding a new constant hardening contribution term that is called in the stateful property initialization method. An example of this second approach is used in the `initQpStatefulProperties` method of [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.html).
> 
> Hope this helps, Stephanie

Hi @sapitts. Thank you for your information. It is helpful. I am wondering if the example model for HCP you mentioned can be used for FCC or BCC structure. If so, what can I make some modifications to this model (https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/hcp_single_crystal/update_method_hcp_aprismatic_capyramidal.i). Appreciate your help.

**Reply by sapitts** [reply]
I'd strongly recommend against using the HCP class for a cubic crystal. As the name suggests, the class I linked to as an example applies assumptions unique to HCP crystal structures. The Kalidindi class that you linked to, source code [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C), is intended for FCC materials.

You will need to write some code to achieve your objective here. I'd advise creating a new class that inherits from `CrystalPlasticityKalidindiUpdate` and overwrite the `initQpStatefulProperties` method to add the new slip resistance contributions

**Reply by xchengood** [reply]
> I'd strongly recommend against using the HCP class for a cubic crystal. As the name suggests, the class I linked to as an example applies assumptions unique to HCP crystal structures. The Kalidindi class that you linked to, source code [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C), is intended for FCC materials.
> 
> You will need to write some code to achieve your objective here. I'd advise creating a new class that inherits from `CrystalPlasticityKalidindiUpdate` and overwrite the `initQpStatefulProperties` method to add the new slip resistance contributions

Hi @sapitts. Very appreciate your information.  The method of CrystalPlasticityHCPDislocationSlipBeyerleinUpdate you mentioned is developed based on the paper (https://www.sciencedirect.com/science/article/pii/S074964190700109X). My questions are below.

1. Does that method for HCP take the the VPSC model menthoned in the paper into consideration? If not, could you share any ideas about how to combine VPSC model to that method with me? So that I can apply the model to polycrystal.

2. How can I get and output pole figures presented in the paper in Moose?

3. I see there are many variables referring to slip resistance and forest dislocation and substructure dislocation in the program (https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/hcp_single_crystal/update_method_hcp_aprismatic_active.i). Could you explain more about those variables in Postprocessors module? I also want to study the slip resistance, forest and substructure dislolcation change with time or temperature. But I am confused about the specific meaning of each variable.

Thanks a lot.

**Reply by ngrilli** [reply]
Dear @xchengood 

I have implemented some time ago the Beyerlein, Tome model in MOOSE:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityUranium.C
This is based on forest and substructure dislocation densities as state variables.

VPSC is not really a constitutive model but it is a type of solver, different from the finite element method in MOOSE.
VPSC means that each grain is considered immersed into a uniform medium and treated as Eshelby inclusion to calculate the plastic deformation. It's different from CPFEM.

Converting a pole figure from a paper into grain orientations is generally complicated,
you could parse the colours in the figure, convert into a distribution function, then use monte carlo methods
to pick grain orientations.
It's better if you ask to the corresponding author if they can provide the original Euler angles data. 

Best Regards,
Nicolò

**Reply by sapitts** [reply]
Hi @xchengood,

1.  As @ngrilli describe, the crystal plasticity FEM implementation in MOOSE is different from the VPSC implementation described in Beyerlein and Tome, although both can use the same constitutive models. An overview of the crystal plasticity algorithm is given on the documentation page for [ComputeMultipleCrystalPlasticityStress](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html). There are many papers by Tome and collaborators to describe VPSC.
2. We do not currently have the capability to generate pole figures in MOOSE. We certainly do accept pull requests though!
3. The slip resistance, forest dislocations, and substructure density are explained in the documentation page for [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.html) as well as the papers cited on that documentation page. The numbers appended to the slip resistance and forest dislocations in the input file correspond to specific slip systems, in the order given in the text file [hcp_aprismatic_capyramidal_slip_sys.txt](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/hcp_single_crystal/hcp_aprismatic_capyramidal_slip_sys.txt).

If you'd like to discuss this HCP class and/or the forest dislocations and slip resistances further, please open a new discussion post specific to this new topic. In that way we will be best following the discussion board guidelines.  Thanks!

**Reply by sapitts** [reply]
Hi @ngrilli, 
It'd be great if you would be open to making a pull request to MOOSE for your crystal plasticity code, after generalizing the material and switching to the latest base class version. @dewenyushu has performed some bench-marking which shows that the `ComputeMultipleCrystalPlasticityStress` and `CrystalPlasticityStressUpdateBase` version is faster. We've also fixed some bugs from the older `FiniteStrainCrystalPlasticity` and `UserObject` versions. We'd be glad to assist you in navigating the documentation and testing requirements for a MOOSE pull request.

It would be nice to have contributed code from our community of crystal plasticity users in the MOOSE repository!

Best Regards,
Stephanie

**Reply by ngrilli** [reply]
Hi @sapitts 

I have only recently learned the User Object Crystal Plasticity code,
therefore my previous code is mainly done using the monolithic version.

I think my code for the Beyerlein, Tomé model is very similar to CrystalPlasticityHCPDislocationSlipBeyerleinUpdate

I have other classes that may be interesting to the MOOSE CP community,
I have recently coupled CP and phase field fracture:

https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainUObasedCPDamageVol.C

my understanding is that there is no way to do that with existing classes.

Another interesting development is the implementation of continuum dislocation dynamics with transport equations,
the model by Thomas Hochrainer, Stefan Sandfeld, Michael Zaiser, Peter Gumbsch:

https://www.sciencedirect.com/science/article/pii/S0022509613001877

coupled with the CP model using Orowan's law, but it's still using the monolithic code:

https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityDislo.C

It works but the code is poorly written and need to be improved.

I did the pull request procedure once about 5 years ago and I may still remember how to do.

@DaijunHU is working with me on this
and we will be happy to contribute in the future.
We will let you know once we are ready.

Best Regards,
Nicolò

**Reply by xchengood** [reply]
Hi @sapitts @ngrilli . Thank you very much for your information. I really learned a lot. But I still have some confusion.

1. As you mentioned 'the VPSC is a kind of solver and it is a constitutive model', is there VPSC solver in Moose? If so, how to use it and are there any examples? If not, How can I apply the HCP model to polycrystal, not only for single crystal?

2. I also want to study how the stress-strain curve change with temperature. I add the the following block to the [AuxKernels]
  [temperature]
    type = FunctionAux
    variable = temperature
    function = '300+400*t' # temperature increases at a constant rate
    execute_on = timestep_begin
  []
Is my method correct? Basically, I got stress-strain curve at 300K and 1600K. The yield strength shows significant decrease, but the Young's modulus does not show any change. Two curves have the same slope. This does not make sense. If my method is not correct, could you share any ideas with me? Basically, the Young's modulus should decrease with the increase of temperature.

Thank you. Best wishes!

**Reply by sapitts** [reply]
Hi @xchengood,
Please make a new discussion post to discuss these new issues. A new discussion post, with a title to capture these questions about including temperature effects and multiple crystals , will help other users find our discussion in the future.

Thanks,
Stephanie

**Reply by xchengood** [reply]
> Hi @xchengood, Please make a new discussion post to discuss these new issues. A new discussion post, with a title to capture these questions about including temperature effects and multiple crystals , will help other users find our discussion in the future.
> 
> Thanks, Stephanie

Hi @sapitts Thank you. I will make a new discussion post.

**Comment by dewenyushu** [comment]
> How can I get and output pole figures presented in the paper in Moose?

We had some issues calculating the updated Euler angles. I am working on a PR (#19494) to fix this issue and enable the capability to output updated Euler angles. 

To visualize the texture, you will need to output the updated Euler angles in .csv files via appropriate postprocessors, and then use additional tools, e.g., MTEX, to plot the pole figures.
